R/generate.R

```diff
@@ -0,0 +1,55 @@
+int_types <- c("integer", "smallint")
+logi_types <- c("boolean")
+num_types <- c("date", "numeric", "real")
+date_types <- c("timestamp without time zone", "timestamp with time zone")
+
+generate_type_check <- function(column_name, data_type, is_nullable) {
+  r_type <- ifelse(data_type %in% int_types, "\"integer\"",
+                   ifelse(data_type %in% logi_types, "\"logical\"",
+                          ifelse(data_type %in% num_types, "\"double\"", "\"character\"")))
+
+  type_check <- paste0("typeof(", column_name, ") == ", r_type)
+  type_check <- ifelse(data_type %in% date_types, paste0("inherits(", column_name, ", \"POSIXct\")"), type_check)
+
+  na_check <- ifelse(is_nullable == "YES", paste0("is.na(", column_name, ") || "), "")
+  paste0("stopifnot(", na_check, type_check, ")")
+}
+
+generate_fake_table_function <- function(table) {
+  ret <- paste0("fake_", table$table_name[[1]], " <- function(")
+  args <- paste0(table$column_name, ifelse(table$is_nullable == "YES", " = NA", ""), collapse = ", ")
+  dat <- paste(table$column_name, "=", table$column_name, collapse = ", ")
+  ret <- paste0(ret, args, ") {\n ")
+  type_checks <- paste0(table$type_check, collapse = "\n  ")
+  ret <- paste(ret, type_checks)
+  ret <- paste0(ret, "\n  data.frame(", dat, ")\n}")
```

Comment:
A pattern I've found very useful if generating code is to try and assemble statements, then write something to build them into blocks with indenting etc, to try and keep the logic and the formatting more separate. It does not always work.  See for example https://github.com/mrc-ide/odin/blob/master/R/generate_js_util.R#L5-L16